# List comprehensions-2

Consider some extensions to the basic list comprehension syntax, allowing to use it for much more sophisticated tasks. 

## Example. 
Count the number of integer points (namely points with integer coordinates) inside a circle of radius 20

<img src='ListComprehensions2_circle.png' style='width:600px;height:600px;float:left'>

In [1]:
r=20 #specify the radius

In [2]:
#note that x,y inside the circle need to satisfy a condition x**2+y**2<=r**2
#with that they can range within -r..r each

In [3]:
#traditional loop
points=[]
for x in range(-r,r+1): #consider all possible x,y
    for y in range(-r,r+1):
        if x**2+y**2<=r**2: #make sure the point is inside the circle
            points+=[(x,y)] #add the point

In [4]:
len(points) #number of points

1257

In [5]:
points=[(x,y) for x in range(-r,r+1) for y in range(-r,r+1) if x**2+y**2<=r**2] #list comprehension

In [6]:
len(points)

1257

So we got the answer. Does this number actually tell us anything? Can we check if this number is correct?

Notice that the number of integer points inside a shape can serve as a Monte-Carlo approximation of the shape's area since each of such integer points basically covers a unit area. Most of those area are completely inside the shape, some of those near the borderline - only partly, but those are minority. So the number of points approximate the number of unit areas inside the shape, which is the shape area. 

Now estimate the true area of the circle using a well know formulae - pi
R**2

In [7]:
import math
math.pi*r**2

1256.6370614359173

So with 1257 points, we are actually as close as we could be with an integer approximation

Now getting back to the list comprehensions, we just learned how to construct complex lists using nested loops and conditional statements inside the list definition.